ruby_oop/tic_tac_toe/game.rb

```diff
@@ -0,0 +1,80 @@
+
+# class to set up the game
+class Game
+  attr_accessor :board, :player1, :player2
+
+  def initialize
+    @player1 = create_player
+    @player2 = create_player(ordinal_prompt: 'second')
+    @board = Board.new
+    board.display_board
+  end
+
+  def swap_players
+    turn_number = 1
+    until game_over?
+      if turn_number.odd?
+        puts "#{player1.name.capitalize} is selecting a spot..."
+        @board.board_update(player1.marker)
+        turn_number += 1
+      else
+        puts "#{player2.name.capitalize} is selecting a spot..."
+        @board.board_update(player2.marker)
+        turn_number += 1
+      end
+    end
+  end
+
+  private
+
+  def create_player(ordinal_prompt: 'first')
+    puts "Enter #{ordinal_prompt} player name: "
+    player_name = gets.chomp.capitalize
+    puts 'Enter marker (X or O): '
+    marker = gets.chomp.upcase
+    until marker == 'X' || marker == 'O'
+      puts 'Enter X or O only: '
+      marker = gets.chomp.upcase
+    end
+    Player.new(player_name, marker)
+  end
+
+  # I think this method can be private, but not sure why just yet..
+  def game_over?
+    game_over = false
+    # tried having this ugly long if in a constant, but couldn't make it work.. 
+    if (@board.board[0] == 'X' && @board.board[1] == 'X' && @board.board[2] == 'X') || # first row horizontal
+       (@board.board[0] == 'O' && @board.board[1] == 'O' && @board.board[2] == 'O') || # first row horizontal
+       (@board.board[0] == 'X' && @board.board[3] == 'X' && @board.board[6] == 'X') || # first row vertical
+       (@board.board[0] == 'O' && @board.board[3] == 'O' && @board.board[6] == 'O') || # first row vertical
+       (@board.board[0] == 'X' && @board.board[4] == 'X' && @board.board[8] == 'X') || # top left diagonal
+       (@board.board[0] == 'O' && @board.board[4] == 'O' && @board.board[8] == 'O') || # top left diagonal
+       (@board.board[3] == 'X' && @board.board[4] == 'X' && @board.board[5] == 'X') || # second row horizontal
+       (@board.board[3] == 'O' && @board.board[4] == 'O' && @board.board[5] == 'O') || # second row horizontal
+       (@board.board[6] == 'X' && @board.board[7] == 'X' && @board.board[8] == 'X') || # third row horizontal
+       (@board.board[6] == 'O' && @board.board[7] == 'O' && @board.board[8] == 'O') || # third row horizontal
+       (@board.board[1] == 'X' && @board.board[4] == 'X' && @board.board[7] == 'X') || # second row vertical
+       (@board.board[1] == 'O' && @board.board[4] == 'O' && @board.board[7] == 'O') || # second row vertical
+       (@board.board[2] == 'X' && @board.board[5] == 'X' && @board.board[8] == 'X') || # third row vertical
+       (@board.board[2] == 'O' && @board.board[5] == 'O' && @board.board[8] == 'O') || # third row vertical
+       (@board.board[2] == 'X' && @board.board[4] == 'X' && @board.board[6] == 'X') || # top right diagonal
+       (@board.board[2] == 'O' && @board.board[4] == 'O' && @board.board[6] == 'O') # top right diagonal
```

Comment:
Brute force is perfectly valid when you're figuring things out. 😀

Here are the steps I'd take to refactor when I see some obviously un-DRY.

Look for the actual _variables_ and try to factor them out, i.e. what's different from case to case. In this case I'm seeing two: `X vs O` and the indices of the winning vector.

I'll start with `X v O` because that seems more obvious. Do we care that these are specifically `X` or `O`? No. We really just care that they're all the same (and not `nil`).

```ruby
if (@board.board[0] == @board.board[1] && @board.board[0] == @board.board[2] && !@board.board[0].nil?) || # first row horizontal
```

This conditional works for both `X` and `O`. We don't have to compare _every_ position against every other position because of the _transitive properties of equality™_. (Also `board` is such a weird word when it's written out so many times o_0)

With that out of the way, we can start thinking about the winning vectors. Your instinct about using a constant is good. These appear to be just sets of 3 values `[0, 1, 2], [0, 3, 6]...` so let's throw those into an array:

```ruby
WINNING_VECTORS = [
  [0,1,2], # first row horizontal
  [0,3,6], # first row vertical
  # etc...
]
```

By looping through those, we can test each vector:

```ruby
if WINNING_VECTORS.any? { |v| @board.board[v[0]] == @board.board[v[1]] && @board.board[v[0]] == @board.board[v[2]] && @board.board[v[0]].nil? }
  game_over = true
  puts "Game over!'
elsif # tie condition, etc
```

There's still more refactoring that can happen, but holy crap I already got carried away.

**TLDR**: I'm down to pair on this if need-be. ¯\\\_(ツ)_/¯

Reply:
This is a good refactor. I also think it's good to have gone through the brute force approach to see the value in the refactor in the end. I remember going through something similar in my very first CS course in college, where my prof. basically gave me the same refactoring advice.

I recommend working through this refactor before getting back to the next assignment, as it's important.